pkg/api/api.go

```diff
@@ -130,18 +152,40 @@ func (dvol *DvolAPI) CreateBranch(volumeName, branchName string) error {
 	return dvol.dl.CreateVariant(volumeName, branchName)
 }
 
-func (dvol *DvolAPI) SwitchBranch(volumeName, branchName string) error {
-	return nil
+var NotFound = errors.New("Item not found")
+
+func indexOfVariant(variantName string, variants []string) (int, error) {
+	for idx, variant := range variants {
+		if variant == variantName {
+			return idx, nil
+		}
+	}
+	return -1, NotFound
 }
 
-func (dvol *DvolAPI) CheckoutBranch(volumeName, branchName string, create bool) error {
-	// Get the path to the volume and the branch
-	// If we were asked to create it:
-	// NEEDS TO BE DONE BY DATALAYER CREATEVARIANT
-	// 	Get the HEAD, return an error regarding needing to commit if the HEAD doesn't exist yet
-	// 	Copy the metadata from the active branch to the new branch
-	// 	Copy the head commit (commits/deadbeef1234etc) into the new branch path
-	// Switch to it
+func (dvol *DvolAPI) CheckoutBranch(volumeName, firstBranch, secondBranch string, create bool) error {
+	if create {
+		allVariants, err := dvol.dl.AllVariants(volumeName)
+		if err != nil {
+			return err
+		}
+		if _, err := indexOfVariant(secondBranch, allVariants); err == nil {
+			return fmt.Errorf("Cannot create existing branch %s", secondBranch)
+		}
+		if err := dvol.dl.CreateVariantFromVariant(volumeName, firstBranch, secondBranch); err != nil {
+			return err
+		}
+	} else {
+		branchPath := dvol.dl.VariantPath(volumeName, secondBranch)
+		if _, err := os.Stat(branchPath); err != nil {
```

Comment:
I must have been seeing things while writing the test then - anyway you're
right though - it does work.

On Thu, 17 Mar 2016 at 11:44 Bridget McErlean <notifications@github.com>
wrote:

> In pkg/api/api.go
> <https://github.com/ClusterHQ/dvol/pull/56#discussion_r56491156>:
>
> > -	// Switch to it
> > +func (dvol *DvolAPI) CheckoutBranch(volumeName, firstBranch, secondBranch string, create bool) error {
> > +	if create {
> > +		allVariants, err := dvol.dl.AllVariants(volumeName)
> > +		if err != nil {
> > +			return err
> > +		}
> > +		if _, err := indexOfVariant(secondBranch, allVariants); err == nil {
> > +			return fmt.Errorf("Cannot create existing branch %s", secondBranch)
> > +		}
> > +		if err := dvol.dl.CreateVariantFromVariant(volumeName, firstBranch, secondBranch); err != nil {
> > +			return err
> > +		}
> > +	} else {
> > +		branchPath := dvol.dl.VariantPath(volumeName, secondBranch)
> > +		if _, err := os.Stat(branchPath); err != nil {
>
> VariantPath doesn't need to be exported for tests in datalayer_test.go as
> it is part of the datalayer package.
>
> My thought here was that if we use the same check (branchExists which
> checks the result of datalayer.AllVariants) for checking the existence of
> a variant both in the create and non-create cases, then we don't need to
> call VariantPath which means we don't need to export it.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/ClusterHQ/dvol/pull/56/files/6bc9ed8a9100027f854dd2c79a1c1c44867e9b0e#r56491156>
>
-- 

Brendan Donegan - 07771376827
Lead QA Engineer @ ClusterHQ